### DISCUSSION TITLE
Refinement (custom initial MOOSE mesh)

### DISCUSSION THREAD

**Discussion by mangerij** [idaholab/moose#28272] [Q&A Meshing] [answered]
Hi,

Is this AMR scheme possible (attached)?

Note that the dashed blue line separates two blocks.
At t = 0, the mesh needs to look like c)

How would I make a) with MOOSE mesh? is it possible to do a uniform refinement in the top block only to get to c) and still preserve the coarsening levels?

as the simulation evolves, we would like to be able to move between the four levels.

thanks

![IMG-20240731-WA0009(1)](https://github.com/user-attachments/assets/e1c8f667-41ab-4a0a-8c2a-877451d16268)

**Comment by GiudGiud** [comment]
Hello

It's not possible with all discretizations. There are more than two levels of refinement  difference between two neighboring cells

We support one level of refinement. Beyond that you will need to add an interface kernel and implement a special treatment 

Guillaume

**Reply by mangerij** [reply]
I was very careful to ensure there is only one level of refinement between neighboring cells here.

**Reply by GiudGiud** [reply]
I must not have zoomed enough on d)

This would be supported then

**Reply by GiudGiud** [reply]
Though it is only supported when libmesh performs this refinement. 

If this is done in an external mesher and imported in exodus, it s not natively supported

**Reply by GiudGiud** [reply]
We would need to add the refinement flags to each element when it's read and possibly create the parent elements too

**Reply by mangerij** [reply]
yeah I only want to do this with the MOOSE mesh but confused on how to make initially a)

and then if uniform_refine is block restrictable

**Reply by GiudGiud** [reply, marked as answer]
You can make initial a) with the RefineSidesetGenerator though you may want to modify it a bit to get single sided refinement 

Then for block restricted refinement use the RefineBlockGenerator

They can be chained in the Mesh block

**Reply by mangerij** [reply]
Beautiful! thanks @GiudGiud  working example:

```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 2
    ny = 2
    nz = 3
    xmin = -10.0
    xmax = 10.0
    ymin = -10.0
    ymax = 10.0
    zmin = -10.0
    zmax = 10.0
    elem_type = HEX8
  []
  [subdomains]
    type = SubdomainBoundingBoxGenerator
    input = gen
    bottom_left = '-10.0 -10.0 -20.0'
    block_id = '1'
    top_right = '10.0 10.0 0.0'
    location = INSIDE
  []
  [film_interface]
    type = SideSetsBetweenSubdomainsGenerator
    input = subdomains
    primary_block = 0
    paired_block = 1
    new_boundary = 52
  []
  [refine_block0]
    type = RefineBlockGenerator
    input = film_interface
    block = '0'
    refinement = 1
  []
  [refine_sub]
    type = RefineSidesetGenerator
    input = refine_block0
    boundaries = '52'
    refinement = 2
    boundary_side = 'secondary'
    enable_neighbor_refinement = false
  []
[]
```

![image](https://github.com/user-attachments/assets/c8a42f9b-3247-4867-87cf-7e216c36f99f)

**Reply by mangerij** [reply]
@GiudGiud one more question: will RefineBlockGenerator rewrite the min h_level for AMR?

**Reply by GiudGiud** [reply]
No you can recover the refined elements with mesh coarsening if needed 

And they can be refined further by AMR as well
